tests/purs/failing/InvalidDerivedInstance3.out

```diff
@@ -0,0 +1,10 @@
+Error found:
+at tests/purs/failing/InvalidDerivedInstance3.purs:10:1 - 10:40 (line 10, column 1 - line 10, column 40)
```

Comment:
Shouldn't this be reported at line 8?

Reply:
Because derived instances are currently desugared before type checking, unfortunately no.

Reply:
@natefaubion Sorry, I don't understand why that's relevant. When `deriveNewtype` finds the type declaration for the newtype, doesn't that have source annotations on it? Can't any errors raised in `replaceAllTypeSynonymsM` be hinted with the position of that type declaration? Or maybe `replaceAllTypeSynonyms'` should do the hinting internally based on the (currently ignored) source annotation of the type it's replacing? I don't see how any of that would depend on type checking happening first.

Reply:
I don't think expanded type synonyms should retain the ranges of the defining declaration. I could be wrong, but I think the issue is happening because it's inlining the type and then expanding it. So its treating the instance head like it was written as:
```
instance newtypeN :: Newtype N S
```

It then tries to expand `S`, which throws, which is happening in the instance declaration. If derived instances were done as part of typechecking (which is actually necessary anyway due to PolyKinds), then this would just be hit when the `data N` declaration is kind checked, which would happen before this instance.

Reply:
Maybe the expansion doesn't have ranges, but the input type to be expanded comes directly from the newtype declaration so the ranges are present in that—unless there's some case I'm missing where that wouldn't be true? Take a look at #3951; it seems to work out fine, at least for the test added here.